util/java/libtiled-java/src/main/java/org/mapeditor/util/StreamHelper.java

```diff
@@ -0,0 +1,140 @@
+	private static final int GZIP_EXTENSION_LENGTH = GZIP_EXTENSION.length();
+	
+    private StreamHelper() {
+    }
+
+	/**
+	 * Opens an {@link InputStream} for reading from the specified location,
+	 * automatically uncompressing data in the GZIP file format if required.
+	 * @param location The filename, path or URL to read
+	 * @return the input stream for reading from the specified location,
+	 * or {@code null} if the location is neither {@link #isUrl(String)} nor {@link #isPathname(String)}
+	 */
+	public static InputStream openStream(String location) throws IOException {
+
+		// (sanity check)
+		if ((location == null) || location.length() < 0) {
+			return null;
+		}
+
+		boolean isUrl = (location.indexOf("://") > 0)
+				|| location.startsWith("file:");
+
+		InputStream in = isUrl
+				? new URL(location).openStream()
+				: new FileInputStream(location);
+
+		return isGzip(location)
+				? ungzip(in)
+				: in;
+	}
+
+	/**
+	 * Opens a connection to the {@code URL} and returns
+	 * an {@link InputStream} for reading from that connection,
+	 * automatically uncompressing data in the GZIP file format if required.
+	 * @param url the URL
+	 * @return the input stream for reading from the URL connetion,
+	 * or {@code null} if the url is {@code null}
+	 */
+	public static InputStream openStream(URL url) throws IOException {
+
+		// (sanity check)
+		if (url == null) {
+			return null;
+		}
+
+		InputStream in = url.openStream();
+		return isGzip(url.getPath()) ? ungzip(in) : in;
+	}
+
+	/**
+	 * @param location The filename, path or URL to check
+	 * @return {@code true} if the filename, path or URL has GZIP extension (ignoring case),
+	 * {@code false} otherwise
+	 */
+	public static boolean isGzip(String location) {
+
+		if (location == null || location.length() < GZIP_EXTENSION_LENGTH) {
+			return false;
+		}
+		final int offset = location.length() - GZIP_EXTENSION.length();
+        return location.regionMatches(true, offset, GZIP_EXTENSION, 0, GZIP_EXTENSION_LENGTH);
+	}
+
+	/**
+	 * @param in the {@link InputStream} to wrap with a {@link GZIPInputStream}
+	 * @return a {@link GZIPInputStream} wrapping the {@code in},
+	 * the same {@code in} if it already was a {@link GZIPInputStream},
+	 * or {@code null} if {@code in} was {@code null}
+	 */
+	public static GZIPInputStream ungzip(InputStream in) throws IOException {
+
+		// (sanity check)
+		if (in == null) {
+			return null;
+		}
+
+		return (in instanceof GZIPInputStream)
+				? GZIPInputStream.class.cast(in)
+				: new GZIPInputStream(in);
+	}
+
+	/**
+	 * @param in the {@link InputStream} to wrap with a {@link BufferedInputStream}
+	 * @return a {@link BufferedInputStream} wrapping the {@code in},
+	 * the same {@code in} if it already was a {@link BufferedInputStream},
+	 * or {@code null} if {@code in} was {@code null}
+	 */
+	public static BufferedInputStream buffered(InputStream in) {
+
+		// (sanity check)
+		if (in == null) {
```

Comment:
Let's not do this additional check. The function will return `null` in any case, since `in instanceof BufferedInputStream` would be `false` and `BufferedInputStream.class.cast(in)` will handle the `null` case fine.

Reply:
Actually, it is the other way around: `new GZIPInputStream(in)` will cause a `NullPointerException`.
I can change these sanity checks by `Objects.requireNonNull(...)`, if you prefer. That would change the behaviour to throw a `NullPointerException`.

Reply:
You're right, let's just leave them be. I don't prefer `Objects.requireNonNull` if the behavior would then be the same as without.